### PR TITLE
build-support/php: init new builder `php.buildComposerWithPlugin`, refactor `php.packages.composer`, etc

### PR DESCRIPTION
**Pull request by drupol**
Refactor `composer` to remove the use of `buildComposerProject` in it, so it doesn't use `composer-local-repo-plugin`, leaving us a total freedom to improve how derivation like `composer-local-repo-plugin` are handled.

This was a necessary evil to make `php.buildComposerWithPlugin` works correctly, or else we end up in infinite recursion due to the fact that `composer-local-repo-plugin` require `composer` which requires `composer-local-repo-plugin`... We should have done that from the beginning.

The Composer plugin `composer-local-repo-plugin` is now built using `php.buildComposerWithPlugin`.

This PR doesn't update the `vendorHash`, it's just some refactoring and cleaning.

### CODE REVIEW

**Comment by drupol** [comment]
@ofborg build php.packages.composer phpunit

**Comment by drupol** [comment]
@ofborg build php.packages.composer phpunit

**Comment by drupol** [comment]
@ofborg build php.packages.composer phpunit php.packages.composer-local-repo-plugin php.packages.cyclonedx-php-composer

**Comment by lucasew** [comment]
Result of `nixpkgs-review pr 312231` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>74 packages built:</summary>
  <ul>
    <li>adminer</li>
    <li>bookstack</li>
    <li>composer-require-checker</li>
    <li>davis</li>
    <li>firefly-iii</li>
    <li>grocy</li>
    <li>librenms</li>
    <li>movim</li>
    <li>n98-magerun</li>
    <li>n98-magerun2</li>
    <li>paratest</li>
    <li>pdepend</li>
    <li>pest</li>
    <li>phel</li>
    <li>php81Packages.box (php82Packages.box ,php83Packages.box)</li>
    <li>php81Packages.castor</li>
    <li>php81Packages.composer</li>
    <li>php81Packages.composer-local-repo-plugin</li>
    <li>php81Packages.cyclonedx-php-composer</li>
    <li>php81Packages.deployer</li>
    <li>php81Packages.grumphp</li>
    <li>php81Packages.phan</li>
    <li>php81Packages.phing</li>
    <li>php81Packages.phive</li>
    <li>php81Packages.php-codesniffer</li>
    <li>php81Packages.php-cs-fixer</li>
    <li>php81Packages.php-parallel-lint</li>
    <li>php81Packages.phpinsights</li>
    <li>php81Packages.phpmd</li>
    <li>php81Packages.phpstan</li>
    <li>php81Packages.psalm</li>
    <li>php81Packages.psysh</li>
    <li>php82Packages.castor</li>
    <li>php82Packages.composer</li>
    <li>php82Packages.composer-local-repo-plugin</li>
    <li>php82Packages.cyclonedx-php-composer</li>
    <li>php82Packages.deployer</li>
    <li>php82Packages.grumphp</li>
    <li>php82Packages.phan</li>
    <li>php82Packages.phing</li>
    <li>php82Packages.phive</li>
    <li>php82Packages.php-codesniffer</li>
    <li>php82Packages.php-cs-fixer</li>
    <li>php82Packages.php-parallel-lint</li>
    <li>php82Packages.phpinsights</li>
    <li>php82Packages.phpmd</li>
    <li>php82Packages.phpstan</li>
    <li>php82Packages.psalm</li>
    <li>php82Packages.psysh</li>
    <li>php83Packages.castor</li>
    <li>php83Packages.composer</li>
    <li>php83Packages.composer-local-repo-plugin</li>
    <li>php83Packages.cyclonedx-php-composer</li>
    <li>php83Packages.deployer</li>
    <li>php83Packages.grumphp</li>
    <li>php83Packages.phan</li>
    <li>php83Packages.phing</li>
    <li>php83Packages.phive</li>
    <li>php83Packages.php-codesniffer</li>
    <li>php83Packages.php-cs-fixer</li>
    <li>php83Packages.php-parallel-lint</li>
    <li>php83Packages.phpinsights</li>
    <li>php83Packages.phpmd</li>
    <li>php83Packages.phpstan</li>
    <li>php83Packages.psalm</li>
    <li>php83Packages.psysh</li>
    <li>phpactor</li>
    <li>phpdocumentor</li>
    <li>phpunit</li>
    <li>pixelfed</li>
    <li>platformsh</li>
    <li>robo</li>
    <li>snipe-it</li>
    <li>vimPlugins.phpactor</li>
  </ul>
</details>

**Comment by drupol** [comment]
@ofborg build php.packages.composer phpunit php.packages.composer-local-repo-plugin php.packages.cyclonedx-php-composer

**Comment by jtojnar** [comment]
> This was a necessary evil to make `php.buildComposerWithPlugin`

How would one go about using multiple plug-ins?

> This PR doesn't update the `vendorHash`, it's just some refactoring and cleaning.

Do you mean the updating is not necessary, or that you did not bother with it?

**Comment by drupol** [comment]
> > This was a necessary evil to make `php.buildComposerWithPlugin`
> 
> How would one go about using multiple plug-ins?

This is not yet supported. I guess it can be done later.
 
> > This PR doesn't update the `vendorHash`, it's just some refactoring and cleaning.
> 
> Do you mean the updating is not necessary, or that you did not bother with it?

It's indeed not necessary.

File: pkgs/development/php-packages/composer-local-repo-plugin/default.nix [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+  src = fetchFromGitHub {
+    owner = "nix-community";
+    repo = "composer-local-repo-plugin";
+    rev = version;
+    hash = "sha256-edbn07r/Uc1g0qOuVBZBs6N1bMN5kIfA1b4FCufdw5M=";
+  };
+
+  composerLock = ./composer.lock;
+  vendorHash = "sha256-SL3HiYTVaUwcEfnRO932MWgOP1VRkxTl3lxLbW0qiTY=";
```
[the remark this answers is not in the document]
Comment (drupol) [reply pkgs/development/php-packages/composer-local-repo-plugin/default.nix:22]: True, fixed.

File: pkgs/development/php-packages/cyclonedx-php-composer/default.nix [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+
+  src = fetchFromGitHub {
+    owner = "CycloneDX";
+    repo = "cyclonedx-php-composer";
+    rev = "v${version}";
+    hash = "sha256-0fb1QiuVJqcB7CAEyB0y60/O9iiibT06mccZYe52dFQ=";
+  };
+
+  composerLock = ./composer.lock;
+  vendorHash = "sha256-QPlHWXXksetNSsv3olmCtPA/VsFVPV09rYQEsPezZoE=";
```
[the remark this answers is not in the document]
Comment (drupol) [reply pkgs/development/php-packages/cyclonedx-php-composer/default.nix:22]: True, fixed.

**Review by patka-123 — APPROVED** [review]

**Review by patka-123 — APPROVED** [review]

File: pkgs/build-support/php/builders/v1/build-composer-with-plugin.nix [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+              versions = {
+                "${pluginName}" = "${pluginVersion}";
+              };
+            };
+          }
+        ];
+      }
+    );
+
+  buildComposerWithPluginOverride =
```
[the remark this answers is not in the document]
Comment (drupol) [reply pkgs/build-support/php/builders/v1/build-composer-with-plugin.nix:41]: We could indeed draft something and make a hierarchical list of what we would like to have so we can work on it in the background. Do you already have some ideas in mind?

**Comment by drupol** [comment]
@jtojnar Is it OK for you?

File: pkgs/build-support/php/builders/v1/build-composer-with-plugin.nix (hunk 3 above)
[the remark this answers is not in the document]
Comment (drupol) [reply pkgs/build-support/php/builders/v1/build-composer-with-plugin.nix:41]: I think I will have a clearer view if we could have the different builders merged and then abstract the common things out of them.

**Review by patka-123 — APPROVED** [review]

**Review by patka-123 — APPROVED** [review]

**Comment by drupol** [comment]
I'll wait until the 23rd of May so that other people can have some time to review this. If no objection, I'll merge it on that date.

**Comment by drupol** [comment]
@nixpkgs-merge-bot merge

**Pull request merged**